### ISSUE TITLE
Record keyset

### ISSUE THREAD

**Issue by tensi0n519** [PhantomVl/sccp_manager#61]
Hello,
Is there a way to add a "Record" to a keyset? I am using Cisco 7975's and when I try to create a custom keyset, there is no "Record" keyset. I'd like to allow there to be a Record softkey that the user can press when connected to a call. Thank you.

**Comment by dkgroot** [comment]
@tensi0n519 
There is already support for recording in the chan-sccp softkeysets. Cisco calls this `monitoring` so the softkey button is called `monitor`. It is in the default softkeyset delivered with chan-sccp. I am not sure if any extra support from sccp_manager is required. But @PhantomVl should be able to help with that.

Besides the `monitor` softkey. There is also a feature button called `monitor` for heavy recording users.

(I understand that `monitor` might not be the best name for this, but that's what it's called atm).

Note:
- The softkey `monitor` is a two state deal (either on or off) during a call
- The feature button has three states: 
  - when activated outside of a call it will record all future calls from that moment forth. When then pressed during the call, it will pause/resume recording.
  - when activated during an ongoing call, it will only record that call, and not continue in the next.

@PhantomVl 
```
[sccpdevice]
....
monitor=yes               ; this allows the monitor feature [yes/no] on this device.
```
This column would have to be added to the realtime sccpdevice table, and you can default it to `yes` to make it easy.

Then Either:
add: `monitor` to the softkeysets (in offhook, ringin and connected). Note it is already in the default one provided with chan-sccp, so it you are not replacing it, they are there.
or add a new feature button like:
```
button = feature, record, monitor
```
to the button config

that's all that's needed

Link to gitter chat about this: :point_up: [March 4, 2020 9:52 AM](https://gitter.im/sccp_manager/community?at=5e5f6c4aec7f8746aab059ea)

**Comment by tensi0n519** [comment]
@dkgroot 
I am running sccp_manager and there doesn't appear to me a "monitor" option to add to a custom keyset. I made a 7975 keyset and I'd like to have the monitor option appear when the line is connected.

![keyset](https://user-images.githubusercontent.com/61072954/75941075-8f680480-5e43-11ea-94ba-6a48fc5d4a06.jpg)

I am also running a couple 8945s. Is it possible to setup a monitor keyset on that as well? Thanks.

**Comment by dkgroot** [comment]
@PhantomVl I guess you will have to add the "monitor" key to the keysets still
@tensi0n519 I am sorry i am not able to help directly. Phantom will have to change/update the source and guide you toward testing with it.

**Comment by tensi0n519** [comment]
Thanks @dkgroot for the response. I'll wait on @PhantomVl and see if he's able to assist. :)

**Comment by tensi0n519** [comment]
@dkgroot Any idea what could be causing this issue? When I set the button to monitor, it gives a weird label of "d" or "r"

![IMG_4450](https://user-images.githubusercontent.com/61072954/76133381-be5db200-5fcc-11ea-9044-12f60eeb51b6.jpg)
![IMG_4451](https://user-images.githubusercontent.com/61072954/76133385-c0277580-5fcc-11ea-963c-996613bc6896.jpg)
![8945 record](https://user-images.githubusercontent.com/61072954/76133413-e3eabb80-5fcc-11ea-9387-f64b7afda0ef.jpg)

**Comment by dkgroot** [comment]
@tensi0n519
That should not be happening, it should change the color of the lamp and the icon, but not mess up the label.
I will look into this tomorrow, on my 8945 (assuming that is the phone you are using from the picture, right).

**Comment by tensi0n519** [comment]
@dkgroot 
Yes that is correct. It is a CP-8945. The button lights appear to work as they should, it is just the button label. I'm using the sccp manger to configure the button. I'm not manually editing the config files.

**Comment by dkgroot** [comment]
I can reproduce the issue, although on my phone it switches from "default" to "red" and back.. and it pop's up a message "conversation is being recorded"
It's very funny though, why the label would change. Also nowhere in the source code do we have the word "red" anywhere. So this seems to be one of those weird things 8945 do. I will need to investigate a little further and figure out a way to make the button label say something logical.
I am manually editing the files, so there seems to be no difference there. Just that your label differs slightly from mine.

Note: These 8945's don't behave like other 79xx series phones, in some cases. So we need to figure this out the hard way.

**Comment by tensi0n519** [comment]
Hmmm. What firmware are you running on your 8945? I'm running "sccp 9-3-2-11". Can you also advise what source code files you are looking through? I'll see if I can come up with anything as well.

**Comment by tensi0n519** [comment]
@dkgroot 
Just updated my firmware to 9-4-2sr3-1. The labels have changed to "default" and "red" as you have reported. Very odd.

**Comment by dkgroot** [comment]
Ooh at least that confirms where the labels are coming from.

It's off behaviour because we are actually sending it a label to set on the button,but it seems to ignore that. I think this might be an error in to firmware to be honest. When looking at the phones debug log (via ssh) it is setting the color of the light to "RED", and maybe they made the mistake to couple that to the feature button label as well. It's a little strange and would suggest nobody noticed this at cisco even over multiple firmwares. Or they just don't use the monitor/record feature button that much on cucm.

I have dont some work  on devstate buttons recently that might have some impact on this (and maybe i made a mistake there), so i will backtrack to an older version chan-sccp to see if that clears things up. I will go back to november last year (before the introduction of customizable devstate).

**Comment by dkgroot** [comment]
Checked against november last year, and the situation was the same. So definitely looks like a firmware issue. The questions is now, what can we do about it.

@PhantomVl have you had any chance to look into adding 'monitor' to the softkeysets, or can @tensi0n519 already do this himself via the softkeyset interface ?

**Comment by tensi0n519** [comment]
@dkgroot @PhantomVl 
So I just figured out how to add the monitor softkey to the 7975. Once I created the custom softkey set with sccp manager, I went back into the sccp.conf and added the monitor softkey. I then reloaded the sccp config and reset all of the 7975's. The monitor (Record) softkey is now there. I am not able to add it using the manager though.

In regards to the 8945, I read through the admin manual and it appears as though there is no such softkey as "monitor" or "record". Maybe this is why it cannot be added to the 8945. Could be another firmware issue?

See for yourself if you'd like, page 117: https://www.cisco.com/c/en/us/td/docs/voice_ip_comm/cuipph/8941_8945/10_0/english/admin_guide/P415_BK_AA30369C_00_adminguide-8941-8945-10_0.pdf

**Comment by dkgroot** [comment]
@tensi0n519 Thanks for the heads up about the monitor softkey in the templates (sccp_manager). So that part is covered

The manual does speak about monitoring and recording, but you are right about the missing softkey in the softkey template table. Odd, right? I guess you will have to live with the feature record button for now. At least with the new firmware the label is readable. In the meantime i will try to figure out what we can do about this, and/or if there is something to do. One of these days i really could use some tcpdump from 8945 hooked up to a cucm.

Ad the issue is no longer related sccp_manager, should we create a new issue in chan_sccp, link it to this one and close this ?

**Comment by tensi0n519** [comment]
> Ad the issue is no longer related sccp_manager, should we create a new issue in chan_sccp, link it to this one and close this ?

Yes that sounds good. We can close this and continue the conversation in a new post.

> One of these days i really could use some tcpdump from 8945 hooked up to a cucm.

What about CME? I have a cisco 2951 ISR in my garage that I could probably setup and test. I don't have access to cucm though.

**Comment by dkgroot** [comment]
Fun fact, i used "strings" on the binary firmware files, and it does contain all the necessary softkeys for recording :-) So i guess there is a way, they just forgot to mention how.
Having access to cucm + 8945 might be able to help here. I will have to ask around a little.

CME should basically be able to do the same i guess. not sure which version you would need to support the 8945 hardware. nor do i have any clue how/what to configure, but we can go on that little journey to maybe ?

Can you make the new chan-sccp issue, so it's under your name, i can copy the rest of the information and link it to this issue report.

**Comment by tensi0n519** [comment]
> CME should basically be able to do the same i guess. not sure which version you would need to support the 8945 hardware. nor do i have any clue how/what to configure, but we can go on that little journey to maybe ?

I'll give it a try tomorrow and see if I can get it working with cme.



> 
> 
> Fun fact, i used "strings" on the binary firmware files, and it does contain all the necessary softkeys for recording :-) So i guess there is a way, they just forgot to mention how.
> Having access to cucm + 8945 might be able to help here. I will have to ask around a little.
> 
> CME should basically be able to do the same i guess. not sure which version you would need to support the 8945 hardware. nor do i have any clue how/what to configure, but we can go on that little journey to maybe ?
> 
> Can you make the new chan-sccp issue, so it's under your name, i can copy the rest of the information and link it to this issue report.

Done. https://github.com/chan-sccp/chan-sccp/issues/507

**Comment by dkgroot** [comment]
@tensi0n519 / @PhantomVl Please close the issue here. Currently testing a fix in chan-sccp to remedy the situation.
